proposals/nnnn-partial-consumption.md

```diff
@@ -0,0 +1,260 @@
+Instead we require only that each of its noncopyable fields be consumed at most once on each path.
+Resilient aggregates, however, cannot be partially consumed when used outside their defining resilience domain.
+
+Extending the `Pair` example above, the following becomes legal:
+
+```swift
+func takeUnique(_ elt: consuming Unique) {}
+extension Pair {
+  consuming func passUniques(_ forward: Bool) {
+    if forward {
+      takeUnique(first)
+      takeUnique(second)
+    } else {
+      takeUnique(second)
+      takeUnique(first)
+    }
+  }
+}
+```
+
+The struct `Pair` has two noncopyable fields, `first` and `second`.
+And there are two paths through the function: the paths taken when `forward` is `true` and when it is `false`.
+On both paths, `first` and `second` are both consumed exactly once.
+
+It's not necessary to consume every field on every path, however.
+For example, the following is allowed as well:
+
+```swift
+extension Pair {
+  consuming func passUnique(_ front: Bool) {
+    if front {
+      takeUnique(first)
+    } else {
+      takeUnique(second)
+    }
+  }
+}
+```
+
+Here, only `first` is consumed on the path taken when `front` is `true` and only `second` on that taken when `front` is `false`.
+
+### Field lifetime extension
+
+When a field is _not_ consumed on some path, its destruction is deferred as long as possible.
+Here, that looks like this:
+
+```swift
+extension Pair {
+  consuming func passUnique(_ front: Bool) {
+    if front {
+      takeUnique(first)
+      // second is destroyed
+    } else {
+      takeUnique(second)
+      // first is destroyed
+    }
+  }
+}
+```
+
+Neither `first` nor `second` can be destroyed _after_ the `if`/`else` blocks because that would require a copy.
+
+### Interaction with resilient types<a name="resilient-aggregates"/>
+
+`public` (and `@usableFromInline`) types defined in a library built with library evolution are "resilient".
+Although a resilient struct may--at the time the library is compiled--actually just combine fields together
+
+```swift
+public struct Resilient : ~Copyable {
+  public var unique: Unique
+}
+```
+
+client code cannot rely on that: the struct may evolve.
+For example, the stored property might become computed in a later release of the library:
+
+```swift
+public struct Resilient : ~Copyable {
+  public var unique: Unique {
+    get {...}
+    set {...}
+  }
+}
+```
+
+Partial consumption relies on viewing a struct as just a combination of values, though.
+This view is what permits the refinement of the check that the whole noncopyable aggregate is consumed on each path
+to checking that each noncopyable field of the aggregate is consumed exactly once.
+
+Because, as this example illustrates, the layout of a resilient struct may change,
+partial consumption isn't permitted for a noncopyable resilient value used outside the resilience domain which defines the type.
```

Comment:
I assume this restriction does not apply to `@frozen` types even if they come from another resilience domain?

Reply:
Yes.